### ISSUE TITLE
Nullness Checker needs to handle JDK 8's optional class separately

### ISSUE THREAD

**Issue by kanha95** [typetools/checker-framework#1228]
Consider this example:

```java

import org.checkerframework.checker.nullness.qual.*;
import java.util.Optional;  

class OptionalExample {  
    
    public static void main(String[] args) {  
         
        String str = null;  
         
        Optional<@Nullable String> checkNull = Optional.ofNullable(str);
        
        if(checkNull.isPresent()){  // check for value is present or not  
            //false positive warning
            String lowercaseString = str.toLowerCase();  
            System.out.print(lowercaseString);  
        }else  
            System.out.println("string value is not present");  
    }  
}

```
Output: Error: [dereference.of.nullable] dereference of possibly-null reference str  Line 14

Nullness checker should handle optional class separately as optional class would ensure check for nullness and under no circumstances would execute the statements of 'IF' condition if the string value is null.

**Comment by mernst** [comment]
This does not seem like very likely code.  It's confusing to check one variable (`checkNull`) and then use a different variable (`str`), so that seems like poor style.  I think that a more likely code snippet would be:

```java
    if (checkNull.isPresent()) {
      String lowercaseString = checkNull.get().toLowerCase();
```

The Nullness Checker permits this code.

More generally, the Checker Framework estimates the value of a given variable, but does not estimate relationships between two variables, except for specific ones listed in the manual https://checkerframework.org/manual/#java-expressions-as-arguments .
We could enrich the Checker Framework by making it track more inter-variable relationships, if that is supported by compelling real-world examples.
